build.mill

```diff
@@ -91,69 +91,74 @@ trait ProbablyTestModule extends BaseScalaModule {
   )
 }
 
-// ------------------------ GROUPS ------------------------
+trait RecursivePublishModule { self: BaseScalaModule =>
+  def publishLocal(localIvyRepo: String = null): Command[Unit] = Task.Command {
+    T.traverse(
+      this.moduleDeps.flatMap {
+        case m: RecursivePublishModule => m.moduleDeps
+        case m => Seq(m)
+      }.collect {
+        case m: PublishModule => m
+      }
+    )(module => module.publishLocal().t)()
+    Result.Success(())
+  }
+}
```

Comment:
Do we need to define a similar `publish` method too, to support _real_ publishing?

Also, my assumption is that this will attempt to publish all the modules exactly _once_ because (and only because) it's the direct dependencies of the `soundness` "bundle" modules, which refer to each of the real modules exactly once. That's probably fine: if a module gets forgotten because it's not included as a _direct_ dependency of one of the bundles, then the first time we try to resolve it from a build tool, we will discover that it's not been published.

Reply:
i think yes, we do
this is also correct, it only publishes modules that are included in one of the soundness.* module dependencies
But you want to be able to just publish ALL modules, right?

Reply:
I think it's fine. Would you like me to merge this as-is, or should I hang on for a `publish` implementation?

Reply:
i'll play with it a bit more - i'll give you green light, when im done